File: 01-sorting/py/04-quick-sort.py
def quick_sort(arr, low, high):
    if low < high:
        pivot_index = partition(arr, low, high)
        quick_sort(arr, low, pivot_index - 1)
        quick_sort(arr, pivot_index + 1, high)

def partition(arr, low, high):
    pivot_index = (low + high) // 2
    pivot = arr[pivot_index]

    # put the pivot at the end
    arr[high], arr[pivot_index] = arr[pivot_index], arr[high]

    i = low
    for j in range(low, high):
        if arr[j] < pivot:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1

    # put the pivot at the right position
    arr[i], arr[high] = arr[high], arr[i]
    return i


arr = [4, 2, 7, 3, 6, 2, 5, 1]
n = len(arr)
quick_sort(arr, 0, n - 1)
print("Sorted array is:", arr)

File: 01-sorting/go/04-quick-sort.go
package main

import (
	"fmt"
)

func partition(arr []int) int {
	pivot := arr[len(arr)-1]
	i := 0
	for j := 0; j < len(arr)-1; j++ {
		if arr[j] < pivot {
			arr[i], arr[j] = arr[j], arr[i]
			i++
		}
	}
	arr[i], arr[len(arr)-1] = arr[len(arr)-1], arr[i]
	return i
}

func quicksort(arr []int) {
	if len(arr) < 2 {
		return
	}
	pivotIdx := partition(arr)
	quicksort(arr[:pivotIdx])
	quicksort(arr[pivotIdx+1:])
}

func main() {
	arr := []int{8, 3, 1, 7, 0, 10, 2}
	fmt.Println("Before:", arr)
	quicksort(arr)
	fmt.Println("After: ", arr)
}
